Sources/NIOCore/Codec.swift

```diff
@@ -740,12 +727,77 @@ extension ByteToMessageHandler: RemovableChannelHandler {
     }
 }
 
+/// A protocol for straightforward encoders which encode custom messages to `ByteBuffer`s.
+/// To add a `MessageToByteEncoder` to a `ChannelPipeline`, use
+/// `channel.pipeline.addHandler(MessageToByteHandler(myEncoder)`.
+public protocol MessageToByteEncoder {
+    associatedtype OutboundIn
+
+    /// Called once there is data to encode.
+    ///
+    /// - parameters:
+    ///     - data: The data to encode into a `ByteBuffer`.
+    ///     - out: The `ByteBuffer` into which we want to encode.
+    func encode(data: OutboundIn, out: inout ByteBuffer) throws
+
+    /// Called once there is data to encode and  the `MessageToByteDecoder` is about to leave
+    ///
+    /// - parameters:
+    ///     - data: The data to encode into a `ByteBuffer`.
+    ///     - out: The `ByteBuffer` into which we want to encode.
+    func encodeLast(data: OutboundIn, out: inout ByteBuffer) throws
+
+    /// Called once this `MessageToByteEncoder` is removed from the `ChannelPipeline`.
+    ///
+    /// - parameters:
+    ///     - context: The `ChannelHandlerContext` which this `ByteToMessageDecoder` belongs to.
+    mutating func encoderRemoved(context: ChannelHandlerContext)
+
+    /// Called when this `MessageToByteEncoder` is added to the `ChannelPipeline`.
+    ///
+    /// - parameters:
+    ///     - context: The `ChannelHandlerContext` which this `MessageToByteEncoder` belongs to.
+    mutating func encoderAdded(context: ChannelHandlerContext)
```

Comment:
I don't think either of these methods need the `ChannelHandlerContext`: we don't currently expose this to `MessageToByteEncoder`s anyway.